File: packages/shared-types/src/actions.ts
export type ActionStatus =
  | "pending"
  | "approved"
  | "denied"
  | "running"
  | "completed"
  | "failed";

export type ActionType =
  | "call_contact"
  | "open_guidance"
  | "open_evacuation_page"
  | "search_transport"
  | "approve"
  | "deny";

export interface ActionLog {
  id: string;
  user_id: string;
  action_type: ActionType;
  status: ActionStatus;
  payload_json?: Record<string, unknown>;
  result_json?: Record<string, unknown>;
  requires_approval: boolean;
  approved?: boolean;
  created_at: string;
}

export interface ApprovalCard {
  action_id: string;
  action_type: ActionType;
  label: string;
  description: string;
  requires_approval: true;
}

export interface CallContactPayload {
  contact_id: string;
  message?: string;
  alert_id?: string;
}

export interface OpenGuidancePayload {
  alert_id: string;
  session_id?: string;
}

File: apps/api/app/schemas/actions.py
from typing import Any, Dict, Optional
from uuid import UUID

from pydantic import BaseModel


class CallContactRequest(BaseModel):
    contact_id: UUID
    alert_id: Optional[UUID] = None
    message: Optional[str] = None


class OpenGuidanceRequest(BaseModel):
    alert_id: UUID
    session_id: Optional[str] = None


class ApproveActionRequest(BaseModel):
    action_id: UUID
    reason: Optional[str] = None


class ActionLogResponse(BaseModel):
    id: UUID
    user_id: str
    action_type: str
    status: str
    payload_json: Optional[Dict[str, Any]] = None
    result_json: Optional[Dict[str, Any]] = None
    requires_approval: bool
    approved: Optional[bool] = None
